src/Gaffer/Context.cpp

```diff
@@ -152,6 +156,25 @@ Context::Context( const Context &other, Ownership ownership )
 				break;
 		}
 	}
+
+	if( canceller )
```

Comment:
> comment here about why a canceller cannot be replaced would help me understand

I've added a comment that hopefully explains things in 9cf483d.

> Also what if you attempt to replace a canceller with the same canceller?

Technically that wouldn't cause a problem, but I think it would indicate an unexpected use of the API. The only place we should need to create a context with a canceller is in a UI component about to launch a computation. When a Node generates a copy of a context for use in pulling on its inputs (in a TimeWarp for example) we want to rely on the implicit propagation of the canceller. Propagating the canceller explicitly would indicate that the node was using the heavyweight `Context( other, canceller )` copy constructor, whereas we want the node to use the optimised EditableScope instead.

This is made a little more confusing than necessary by the `Context( other, ownership )` constructor overload. This will be removed in the future, leaving us with just two operations :

1. Copy (deep copy) construct a context, optionally applying a canceller. Used by the UI and anything needing a persistent context.
2. Construct an EditableScope, which is a cheap temporary stack-based copy used by Nodes to generate upstream contexts.

Writing this has made me think that it might be better to have a `CancellerScope` rather than use the copy constructor to make a cancelleable context. That could automatically ensure that the lifetime of the Context didn't extend past the lifetime of the canceller. Does that make sense?

Reply:
> Writing this has made me think that it might be better to have a CancellerScope rather than use the copy constructor to make a cancelleable context. That could automatically ensure that the lifetime of the Context didn't extend past the lifetime of the canceller. Does that make sense?

Thanks for the comment & explaination. I do like the idea of the CancellerScope to manage lifetime.